ev3dev/core.py

```diff
@@ -2615,27 +2610,121 @@ def __init__(self, address=None, name_pattern=SYSTEM_DEVICE_NAME_CONVENTION, nam
     #: Calibration ???
     MODE_IR_CAL = 'IR-CAL'
 
-
     MODES = (
-      'IR-PROX',
-      'IR-SEEK',
-      'IR-REMOTE',
-      'IR-REM-A',
-      'IR-CAL',
+      MODE_IR_PROX,
+      MODE_IR_SEEK,
+      MODE_IR_REMOTE,
+      MODE_IR_REM_A,
+      MODE_IR_CAL
     )
 
+    _BUTTON_VALUES = {
+            0: [],
+            1: ['red_up'],
+            2: ['red_down'],
+            3: ['blue_up'],
+            4: ['blue_down'],
+            5: ['red_up', 'blue_up'],
+            6: ['red_up', 'blue_down'],
+            7: ['red_down', 'blue_up'],
+            8: ['red_down', 'blue_down'],
+            9: ['beacon'],
+            10: ['red_up', 'red_down'],
+            11: ['blue_up', 'blue_down']
+            }
+
+    #: Handles ``Red Up`` events.
+    on_red_up = None
+
+    #: Handles ``Red Down`` events.
+    on_red_down = None
+
+    #: Handles ``Blue Up`` events.
+    on_blue_up = None
+
+    #: Handles ``Blue Down`` events.
+    on_blue_down = None
+
+    #: Handles ``Beacon`` events.
+    on_beacon = None
+
+    def __init__(self, address=None, name_pattern=SYSTEM_DEVICE_NAME_CONVENTION, name_exact=False, **kwargs):
+        super(InfraredSensor, self).__init__(address, name_pattern, name_exact, driver_name=['lego-ev3-ir'], **kwargs)
 
     @property
     def proximity(self):
         """
         A measurement of the distance between the sensor and the remote,
         as a percentage. 100% is approximately 70cm/27in.
         """
+        self.mode = self.MODE_IR_PROX
+        return self.value(0)
 
-        if self.auto_mode:
-            self.mode = self.MODE_IR_PROX
+    # BeaconSeeker
+    def heading(self, channel=1):
+        """
+        Returns heading (-25, 25) to the beacon on the given channel.
+        """
+        self.mode = self.MODE_IR_SEEK
+        channel = max(1, min(4, channel)) - 1
+        return self.value(channel * 2)
+
+    def distance(self, channel=1):
+        """
+        Returns distance (0, 100) to the beacon on the given channel.
+        Returns -128 when beacon is not found.
+        """
+        self.mode = self.MODE_IR_SEEK
+        channel = max(1, min(4, channel)) - 1
+        return self.value((channel * 2) + 1)
+
+    def heading_and_distance(self, channel=1):
+        """
+        Returns heading and distance to the beacon on the given channel as a
+        tuple.
+        """
+        return (self.heading(channel), self.distance(channel))
+
+    # RemoteControl
+    def red_up(self, channel=1):
```

Comment:
`is_red_up_pressed`? or `red_up_pressed`? I originally read this thinking that this was asking whether "red" was released.

Reply:
yeah the up/down naming convention is a little odd.  I don't think we should change this though, we use the RemoteControl a fair amount in the demos, I imagine these are used a lot.

Reply:
Now that we are not worried about being backwards compatible do we want to change the names of `red_up`, `red_down`, etc?  IMO they are pretty confusing. Brainstorming....
- We could label them `top_left`, `bottom_left`, `top_right`, `bottom_right`
- We could rename `red_up()` to `is_top_left_pressed()`, etc 
- We could rename `on_red_up` to `on_top_left_press`, etc